File: ca01/gpt.py
'''
Code for interacting with GPT-3 in Python. 
We created a mock Grammarly using GPT.

To run this you need to 
* first visit openai.com and get an APIkey, 
* which you export into the environment as shown in the shell code below.
* next create a folder and put this file in the folder as gpt.py
* finally run the following commands in that folder

On Mac
% pip3 install openai
% export APIKEY="......."  # in bash
% python3 gpt.py

On Windows:
% pip install openai
% $env:APIKEY="....." # in powershell
% python gpt.py
'''
import openai


class GPT():
    ''' make queries to gpt from a given API '''
    def __init__(self,apikey):
        ''' store the apikey in an instance variable '''
        self.apikey=apikey
        # Set up the OpenAI API client
        openai.api_key = apikey #os.environ.get('APIKEY')

        # Set up the model and prompt
        self.model_engine = "text-davinci-003"

    def getResponse(self,prompt):
        ''' Generate a GPT response '''
        completion = openai.Completion.create(
            engine=self.model_engine,
            prompt=prompt,
            max_tokens=1024,
            n=1,
            stop=None,
            temperature=0.8,
        )

        response = completion.choices[0].text
        return response
    
    def get_summarization(self, text):
        '''Summarize a given text'''
        prompt = "Summarize the following text in one or two sentences" + text
        response = self.getResponse(prompt)
        return response

    def get_translation(self, text):
        '''Translate a given text to Spanish'''
        prompt = f"Translate the following text to Spanish\n\n{text}:"
        response = self.getResponse(prompt)
        return response
 
    def get_paraphrase(self, text):
        '''Paraphrase a given text'''
        prompt = "Paraphrase the following text:" + text
        response = self.getResponse(prompt)
        return response

    def get_poem(self, prompt):
        ''' Generate a GPT poem based on a given prompt '''
        prompt = "Write a poem about:\n\n" + prompt
        response = self.getResponse(prompt)
        return response

    def edit_grammar(self, text):
        '''Generate a GPT response to correct the grammar of a given text.'''
        prompt = "Correct the grammar of the following text:"
        full_prompt = f"{prompt}\n\n{text}"
        response = self.getResponse(full_prompt)
        return response
    
    def get_synonyms(self, word):
        '''Generate synonyms for the word {word}.'''
        prompt = f"Generate synonyms for the word {word}:"
        response = self.getResponse(prompt)
        return response



if __name__=='__main__':
    '''
    '''
    import os
    g = GPT(os.environ.get("APIKEY"))
    print(g.getResponse("what does openai's GPT stand for?"))

File: ca02/routes/gpt.js
const express = require("express");
const router = express.Router();
const path = require('path');
const APIKEY = process.env.APIKEY;
const { Configuration, OpenAIApi } = require("openai");
const configuration = new Configuration({
  apiKey: APIKEY,
});
const openai = new OpenAIApi(configuration);

isLoggedIn = (req,res,next) => {
  if (res.locals.loggedIn) {
    next()
  } else {
    res.redirect('/login')
  }
}

// Getthe response from the GPT API
async function getResponse(text){
  const gptResponse = await openai.createCompletion({
    prompt: text, 
    max_tokens: 64,
    n: 1,
    temperature: 0.5,
    frequency_penalty: 0,
    presence_penalty: 0,
    model: "text-davinci-003",
  })
  return gptResponse.data.choices[0].text;
}

// Route for the summaries page
router.get("/index/summaries", (req, res) => {
  res.render("summaries");
});

// Route to handle the summaries form submissions
router.post("/index/summaries", async (req, res) => {
  const prompt = "Summarize this text: " + req.body.prompt;
  const summary = await getResponse(prompt);
  res.render("summaries", { summary });
});

// Route for the Spanish translator page
router.get("/index/spanishTranslations", (req, res) => {
  res.render("spanishTranslations");
});

// Route to handle Spanish translator form submissions
router.post("/index/spanishTranslations", async (req, res) => {
  const prompt = "Translate this text to Spanish: " + req.body.prompt;
  const translation = await getResponse(prompt);
  res.render("spanishTranslations", { translation });
});

// Route for the poem page
router.get("/index/poems", (req, res) => {
  res.render("poems");
});

// Route to handle poem form submissions
router.post("/index/poems", async (req, res) => {
  const prompt = "write a poem about: " + req.body.prompt;
  const poem = await getResponse(prompt);
  res.render("poems", { poem });

});

// Route for the Spanish translator page
router.get("/index/synonyms", (req, res) => {
  res.render("synonyms");
});

// Route to handle Spanish translator form submissions
router.post("/index/synonyms", async (req, res) => {
  const prompt = "Generate synonyms for the word " + req.body.prompt;
  const synonyms = await getResponse(prompt);
  res.render("synonyms", { synonyms });
});

module.exports = router;
